### DISCUSSION TITLE
Is there a way of adding cumulative DeltaTime?

### DISCUSSION THREAD

**Discussion by SinanKarasu** [orchetect/swift-midi#167] [Q&A] [answered]
What I have in mind is like Logic Pro MIDI Event List.
I wanted to do the Position like they do. It is all SwiftUI Table based.
I can do the rest. This is all MIDIKitSMF
https://support.apple.com/guide/logicpro/event-list-interface-lgcp31f1a05a/mac

**Comment by SinanKarasu** [comment]
OK, ChatGPT says:
How can I in MIDIKitSMF get the cumulative DeltaTime in a track?
To get the cumulative DeltaTime in a track using MIDIKitSMF, you can use the MIDIFileTrackIterator class to iterate through each event in the track and sum up the DeltaTime values.

Here's an example code snippet that demonstrates how to do this:

swift
Copy code
import MIDIKit
import MIDIKitIO
import MIDIKitEvents

// assume `fileURL` is a valid file URL pointing to a MIDI file

do {
    let file = try MIDIFile(url: fileURL)
    let track = file.tracks[0] // assuming you want the first track

    var deltaTimeSum: MIDI.Event.TimeStamp = 0

    let iterator = try MIDIFileTrackIterator(track: track)
    while let event = try iterator.next() {
        deltaTimeSum += event.deltaTime
    }

    print("Cumulative DeltaTime in track: \(deltaTimeSum)")
} catch {
    print("Error: \(error)")
}
In this example, we create a MIDIFile instance from the file URL, and then get the first track from the tracks array. We then create a variable deltaTimeSum to store the cumulative DeltaTime, and initialize it to 0.

We then create a MIDIFileTrackIterator instance to iterate through each event in the track. For each event, we add its DeltaTime value to the deltaTimeSum variable.

After iterating through all the events in the track, we print out the deltaTimeSum variable to get the cumulative DeltaTime.

**Reply by orchetect** [reply]
I think GPT is getting confused. There's no MIDIFileTrackIterator object in MIDIKit.

Do you want to generate a list of positions with bars / beats time format?

The library still needs a little fleshing out to do time conversions for MIDI files. But part of this can be achieved.

I'll reply with some example code in a few minutes.

**Comment by orchetect** [comment, marked as answer]
Load a MIDI file, get the first track, and map all events to their delta time at a given PPQ.

```swift
let midiFile = try MIDIFile(url: fileURL)
guard case .track(let trackOne) = midiFile.chunks.first else { return }

// map delta times using a given PPQ
let ppq = 480
let deltaTimes = trackOne.events.map {
    $0.smfUnwrappedEvent.delta.ticksValue(using: .musical(ticksPerQuarterNote: UInt16(ppq)))
}
```

Then you can find the sum:

```swift
let deltaSum = Int(deltaTimes.reduce(into: 0, +=))
```

Here's the (mostly tested) math to generate Bars - Beats - Division - Ticks for display.

```swift
// does not account for time signature changes or track offset, if any

// parameters
let beatsPerBar = 4 // 4/4 time
let subDivisionsPerBeat = 4 // 16th note subdivisions

// calculation
let ticksPerBar = beatsPerBar * ppq
let bar = deltaSum / ticksPerBar
let barsTicks = bar * ticksPerBar
let beatsTicks = deltaSum - barsTicks
let beat = beatsTicks / ppq
let beatDivisionTicks = beatsTicks - (beat * ppq)
let beatDivision = beatDivisionTicks / (ppq / subDivisionsPerBeat)
let ticksPerDivision = ppq / subDivisionsPerBeat
let remainingTicks = beatsTicks % ticksPerDivision

// final string
let timeString = "\(bar) \(beat) \(beatDivision) \(remainingTicks)"
print(timeString)
```

**Reply by orchetect** [reply]
@SinanKarasu 

If you grab main branch, I pushed some changes that will make this API a little cleaner.

- Added a `tracks` property to MIDIFile so it's easier to access tracks directly in an array without enum case unwrapping.
- Added a `delta` property to the events so use of `smfUnwrappedEvent` is not needed when mapping events.

```swift
let midiFile = try MIDIFile(url: fileURL)
guard let trackOne = midiFile.tracks.first else { return }

// map delta times using a given PPQ
let ppq = 480
let deltaTimes = trackOne.events.map {
    $0.delta.ticksValue(using: .musical(ticksPerQuarterNote: UInt16(ppq)))
}
```

I also added a new struct called `MusicalTimeValue` which expands on the math above to converting total elapsed ticks into a bar-beat-division-ticks and back again, as well as format a string for display. Still works upon a static time signature, but can be expanded to factor in time signature changes in the future.

```swift
let mt = MusicalTimeValue(elapsedTicks: deltaSum, beatsPerBar: 4, divisionsPerBeat: 4, ppq: ppq)

print(mt.bar) // Int
print(mt.beat) // Int
print(mt.beatDivision) // Int
print(mt.ticks) // Int

print(mt.stringValue()) // String, ie: "14 1 2 60"
```
